File: USACO.GUIDE/SILVER/Graph Traversal/CF_1833_E.py
from collections import defaultdict


for _ in range(int(input())):
    n = int(input())
    arr = [0] + list(map(int, input().split()))

    components = []
    visited = [0]*(n+1)
    adj = defaultdict(set)
    #DFS
    for i in range(1, n+1):
        if not visited[i]:
            stack = [i]
            visited[i] = 1
            while stack:
                v = stack.pop()
                adj[v].add(arr[v])
                adj[arr[v]].add(v)
                if not visited[arr[v]]:
                    stack.append(arr[v])
                    visited[arr[v]] = 1

    visited = [0]*(n+1)
    for i in range(1, n+1):
        if not visited[i]:
            stack = [i]
            visited[i] = 1
            components.append([i])
            while stack:
                v = stack.pop()
                for nn in adj[v]:
                    if not visited[nn]:
                        stack.append(nn)
                        visited[nn] = 1
                        components[-1].append(nn)

    ans_min = 0
    ans_max = 0

    extras = 0
    for cmp in components:
        if cmp:
            ans_max += 1
            tt = 0
            for v in cmp:
                if len(adj[v]) != 2:
                    tt += 1
            if extras and tt:
                ans_min += 1
                extras -= 1
                tt -= 1
                extras += tt
            else:
                extras += tt
                
    print(ans_max - ans_min, ans_max)
    # print(components)

File: USACO.GUIDE/SILVER/Graph Traversal/CF_1833_E.cpp
#include <bits/stdc++.h>

using namespace std;

typedef int ll;

int main()
{
    ll t;
    scanf("%d", &t);
    while (t--)
    {
        ll n;
        scanf("%d", &n);
        vector<ll> arr(n);
        vector<vector<ll>> components;
        vector<set<ll>> adj(n);
        // vector<ll> visited(n);
        for (ll i = 0; i < n; i++)
        {
            scanf("%d", &arr[i]);
            arr[i] -= 1;
        }
        for (ll i = 0; i < n; i++)
        {
            adj[i].insert(arr[i]);
            adj[arr[i]].insert(i);
        }

        vector<int> vis(n);
        for (ll i = 0; i < n; i++)
        {
            if (!vis[i])
            {
                vector<ll> ss = {i};
                vis[i] = 1;
                components.push_back({i});
                while (ss.size())
                {
                    ll v = ss.back();
                    ss.pop_back();
                    for (auto nn : adj[v])
                    {
                        if (!vis[nn])
                        {
                            ss.push_back(nn);
                            vis[nn] = 1;
                            components.back().push_back(nn);
                        }
                    }
                }
            }
        }
        ll ans_min = 0, ans_max = 0, extras = 0;
        for (auto cmp : components)
        {
            if (cmp.size())
            {
                ans_max += 1;
                ll tt = 0;
                for (auto v : cmp)
                {
                    if (adj[v].size() < 2)
                    {
                        tt += 1;
                    }
                }
                if (extras > 0 && tt > 0)
                {
                    ans_min += 1;
                    extras -= 1;
                    tt -= 1;
                    extras += tt;
                }
                else
                {
                    extras += tt;
                }
            }
        }
        printf("%d %d\n", ans_max - ans_min, ans_max);
        // print_comp(components);
    }
    return 0;
}